drafts/when-elixirs-performance-becomes-rust-y.md

```diff
@@ -0,0 +1,434 @@
+  end
+
+  defp add_if_prime_number(numbers, 1), do: numbers
+
+  defp add_if_prime_number(numbers, 2) do
+    [2 | numbers]
+  end
+
+  defp add_if_prime_number(numbers, n) do
+    case (numbers |> Enum.any?(fn x -> rem(n, x) == 0 end)) do
+      false -> [n | numbers]
+      _ -> numbers
+    end
+  end
+end
+
+time = Time.utc_now
+RustNif.ElixirPrimes.prime_numbers(Enum.into 1..1000000, []) |> IO.inspect
+IO.puts "Elixir task finished after #{Time.diff Time.utc_now, time} seconds"
+```
+
+Running this, the results on my machine were:
+
+```
+{:ok,
+ [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
+  73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151,
+  157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, ...]}
+Elixir task finished after 2288 seconds
+:ok
+```
+
+Well, that's a bit on the long side. Surely we can do better than that.
+Yes we can. The current implementation prepends each new prime number to the start of the list. This list is reversed in the end. See [list handling](http://erlang.org/doc/efficiency_guide/listHandling.html) as to why we do this.
+
+Most likely, the first divisible number for non prime numbers will reside at the start of the list (the number 2 comes to mind). So let's start our `any?` comparison at the reverse of the list.
+
+```Elixir
+defp add_if_prime_number(numbers, n) do
+  # Enum.reverse() is added
+  case (numbers |> Enum.reverse() |> Enum.any?(fn x -> rem(n, x) == 0 end)) do
+    false -> [n | numbers]
+    _ -> numbers
+  end
+end
+```
+
+After this change, the results are
+
+```
+time = Time.utc_now
+RustNif.ElixirPrimes.prime_numbers(Enum.into 1..1000000, []) |> IO.inspect
+IO.puts "Elixir task finished after #{Time.diff Time.utc_now, time} seconds"
+
+{:ok,
+ [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
+  73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151,
+  157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, ...]}
+Elixir task finished after 328 seconds
+:ok
+```
+
+Well that's a lot better! But it still takes a fair bit of time.
+
+## So what are our alternatives?
+
+### Ports & Nifs
+
+We could resort to ports or NIFs.
+There's a good summary of ports and nifs here: https://spin.atomicobject.com/2015/03/16/elixir-native-interoperability-ports-vs-nifs/
+
+In short:
+
+Ports:
++ Safety
++ Error trapping
++ Flexible communication
++ No external erlang/elixir specific libraries required
+- Communication via STDIN and STDOUT
+
+NIFS:
++ Fast and simple implementationn
++ No context switch required
++ Simple debugging
+- Not very safe
+- NIF-specific C libraries
+- Native implementation can leak into Elixir code
+
+### (Web) service
+
+We could define a separate web service, exposing an API that could handle this, in a language better suited for this. We'd need to add a new service to our stack though, and also account for potential network failures and sending potentially large payloads.
```

Comment:
`We could define a separate web service, exposing an API that could handle this, in a language better suited for this.`
This feels like you are saying that Elixir is not good at web services...

Reply:
I've clarified this section